File: src/main/java/jolkert/discordbotlink/DiscordBotLink.java
package jolkert.discordbotlink;

import com.mojang.logging.LogUtils;
import jolkert.discordbotlink.jda.DiscordBot;
import jolkert.discordbotlink.jda.config.BotConfig;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.server.MinecraftServer;
import org.slf4j.Logger;

import javax.security.auth.login.LoginException;
import java.io.*;
import java.util.Properties;

public class DiscordBotLink implements ModInitializer
{
	public static MinecraftServer Server;
	public static final Logger Logger = LogUtils.getLogger();
	public static DiscordBot Bot;
	
	public static final String DEFAULT_PREFIX = "!";
	
	@Override
	public void onInitialize()
	{
		
		BotConfig botConfig = null;
		try
		{
			botConfig = loadConfig();
		}
		catch (IOException exception)
		{
			Logger.error("COULD NOT LOAD CONFIG FILE!", exception);
		}
		
		
		Bot = null;
		if (botConfig == null)
			Logger.error("COULD NOT LOG IN BOT");
		else
		{
			try
			{
				Bot = new DiscordBot(botConfig);
			}
			catch (LoginException exception)
			{
				Logger.error("COULD NOT LOG IN BOT", exception);
			}
		}
	}
	
	public static BotConfig loadConfig() throws IOException
	{
		String filePath = FabricLoader.getInstance().getConfigDir().toString() + "/discord_bot_link.properties";
		File file = new File(filePath);
		
		BotConfig config;
		
		
		
		if (file.createNewFile())
		{
			Properties properties = new Properties();
			properties.setProperty("token", "");
			properties.setProperty("prefix", DEFAULT_PREFIX);
			PrintWriter writer = new PrintWriter(file);
			properties.store(writer, "Config for DiscordBotLink");
			writer.close();
			
			config = new BotConfig("", DEFAULT_PREFIX);
		}
		else
		{
			InputStream inputStream = new FileInputStream(file);
			Properties properties = new Properties();
			properties.load(inputStream);
			
			String token = properties.getProperty("token", "");
			String prefix = properties.getProperty("prefix", DEFAULT_PREFIX);
			
			inputStream.close();
			config = new BotConfig(token, prefix);
		}
		
		return config;
	}
}

File: src/main/kotlin/io/github/jolkert/discordbotlink/DiscordBotLink.kt
package io.github.jolkert.discordbotlink

import io.github.jolkert.discordbotlink.extension.nameWithDiscriminator
import io.github.jolkert.discordbotlink.extension.pascalToSnake
import io.github.jolkert.discordbotlink.extension.pascalToTitle
import io.github.jolkert.discordbotlink.jda.DiscordBot
import io.github.jolkert.discordbotlink.jda.data.BotConfig
import io.github.jolkert.discordbotlink.jda.data.UserDataManager
import net.dv8tion.jda.api.exceptions.InvalidTokenException
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents
import net.fabricmc.loader.api.FabricLoader
import net.minecraft.server.MinecraftServer
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.PrintWriter
import java.util.*

object DiscordBotLink : ModInitializer
{
	@JvmStatic val LOGGER = LoggerFactory.getLogger(DiscordBotLink::class.simpleName!!.pascalToTitle())
	@JvmStatic val CONFIG_FILE = File("${FabricLoader.getInstance().configDir}/${this::class.simpleName!!.pascalToSnake()}.properties")

	lateinit var Bot: DiscordBot private set
	lateinit var Server: MinecraftServer private set

	val users = UserDataManager("${FabricLoader.getInstance().configDir.toString()}/${DiscordBotLink::class.simpleName}/user_data.json")

	override fun onInitialize()
	{
		LOGGER.info("Loading bot config...")
		ServerLifecycleEvents.SERVER_STARTED.register { Server = it }

		val config: BotConfig? = loadConfig()
		if (config != null)
		{
			try
			{
				Bot = DiscordBot(config)
				LOGGER.info("Logged in bot ${Bot.self.nameWithDiscriminator}!")
			} catch (exception: InvalidTokenException)
			{
				LOGGER.error("Provided bot token was not valid!", exception)
			}
		}

		if (!::Bot.isInitialized)
			LOGGER.error("Unable to start Discord bot. This will almost certainly cause a crash later!")


	}
	private fun loadConfig(): BotConfig?
	{
		try
		{
			if (!CONFIG_FILE.createNewFile())
			{
				BotConfig.fromProperties(Properties().apply { FileInputStream(CONFIG_FILE).also { load(it) }.close() }).let {
					PrintWriter(CONFIG_FILE).also { writer ->
						it.toProperties().store(writer, "Config for ${this::class.simpleName!!.pascalToTitle()}")
					}.close()

					return it
				}
			}
			else
			{
				PrintWriter(CONFIG_FILE).also {
					BotConfig.DEFAULT.toProperties().store(it, "Config for ${this::class.simpleName!!.pascalToTitle()}")
				}.close()

				return BotConfig.DEFAULT
			}
		} catch (exception: IOException)
		{
			LOGGER.error("Could not load bot config file!", exception)
			LOGGER.error("Could not log in bot!")
			return null
		}
	}

	private fun BotConfig.Companion.fromProperties(properties: Properties): BotConfig = properties.let {
		BotConfig(
			it.getProperty(BotConfig::token.name.pascalToSnake(), BotConfig.DEFAULT.token),
			it.getProperty(BotConfig::commandPrefix.name.pascalToSnake(), BotConfig.DEFAULT.commandPrefix),
			it.getProperty(BotConfig::linkChannelId.name.pascalToSnake(), BotConfig.DEFAULT.linkChannelId),
			it.getProperty(BotConfig::kickMessage.name.pascalToSnake(), BotConfig.DEFAULT.kickMessage)
		)
	}
}